crates/uv-resolver/src/marker.rs

```diff
@@ -173,6 +192,177 @@ pub(crate) fn normalize(tree: MarkerTree) -> Option<MarkerTree> {
     }
 }
 
+/// Removes redundant expressions from the tree recursively.
+///
+/// This function does not attempt to flatten or clean trees and may leave it in a denormalized state.
+pub(crate) fn filter_all(tree: &mut MarkerTree) {
+    match tree {
+        MarkerTree::And(trees) => {
+            for subtree in &mut *trees {
+                filter_all(subtree);
+            }
+
+            for conjunct in collect_expressions(trees) {
+                // Filter out redundant disjunctions.
+                trees.retain_mut(|tree| !filter_disjunctions(tree, &conjunct));
+
+                // Filter out redundant conjunctions.
+                for tree in &mut *trees {
+                    filter_conjuncts(tree, &conjunct);
+                }
+            }
+        }
+
+        MarkerTree::Or(trees) => {
+            for subtree in &mut *trees {
+                filter_all(subtree);
+            }
+
+            for disjunct in collect_expressions(trees) {
+                // Filter out redundant conjunctions.
+                trees.retain_mut(|tree| !filter_conjunctions(tree, &disjunct));
+
+                // Filter out redundant disjunctions.
+                for tree in &mut *trees {
+                    filter_disjuncts(tree, &disjunct);
+                }
+            }
+        }
+
+        MarkerTree::Expression(_) => {}
+    }
+}
+
+// Collect all expressions from a tree.
```

Comment:
Scroll-by Nit: `//` -> `///` so it's rustdoc

Reply:
I might add a little more rustdoc for cases like these. As it is, I think the comment is effectively strictly redundant with the name and type signature. I think I'd write something like this:

```rust
/// For a given slice of marker trees, this collects
/// all direct leaf expressions from each marker tree,
/// and skips any marker trees that are themselves
/// not leaves.
```

I might also add some context as to _why_ this routine is useful.